homeassistant/bootstrap.py

```diff
@@ -153,31 +174,38 @@ def _async_setup_component(hass: core.HomeAssistant,
                     None, component.setup, hass, config)
         except Exception:  # pylint: disable=broad-except
             _LOGGER.exception('Error during setup of component %s', domain)
-            _async_persistent_notification(hass, domain, True)
-            return False
+            raise HomeAssistantError()
 
         if result is False:
             _LOGGER.error('component %s failed to initialize', domain)
-            _async_persistent_notification(hass, domain, True)
-            return False
+            raise HomeAssistantError()
         elif result is not True:
             _LOGGER.error('component %s did not return boolean if setup '
                           'was successful. Disabling component.', domain)
-            _async_persistent_notification(hass, domain, True)
             loader.set_component(domain, None)
-            return False
+            raise HomeAssistantError()
 
         hass.config.components.add(component.DOMAIN)
 
         hass.bus.async_fire(
             EVENT_COMPONENT_LOADED, {ATTR_COMPONENT: component.DOMAIN}
         )
 
+        # wait until entities are setup
+        if EV_PLATFORM in setup_events:
```

Comment:
What is the use case of waiting for this to finish?

When the entity component is setup, it is ready to listen for discovery events. 

Is this so that we don't fire EVENT_HOMEASSISTANT_START too soon?